### PR TITLE
Custom command tests

### PR DESCRIPTION
**Pull request by georgecrawford**

### CODE REVIEW

File: test/fixtures/hooks.using.custom.commands.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+export default {
+    capabilities: {
+        browserName: 'chrome'
+    },
+
+    mochaOpts: {
+        timeout: 5000
+    },
+
+    onPrepare: NOOP,
```
Comment (christian-bromann) [inline test/fixtures/hooks.using.custom.commands.js]: we don't need them..if they are undefined should happen

File: test/fixtures/sample4.spec.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
+describe('dummy test', () => {
+    before(() => {
+        browser.addCommand('customWdio', function (a) {
+            browser.pause(1000)
+            return a + 1
+        })
+
+        browser.addCommand('customWdioPromise', function async (a) {
+            return browser.pause(1000)
```
Comment (christian-bromann) [inline test/fixtures/sample4.spec.js]: We seem to have a problem in wdio-sync here:
```
  1) MochaAdapter executes custom commands "before all" hook: callee$1$0:
     Uncaught TypeError: Cannot read property 'then' of undefined
      at Object.async (test/fixtures/sample4.spec.js:9:39)
      at Object.customWdioPromise (node_modules/wdio-sync/build/index.js:157:23)
      at Context.<anonymous> (test/fixtures/sample4.spec.js:52:17)
      at node_modules/wdio-sync/build/index.js:226:24
```

**Comment by christian-bromann** [comment]
Awesome! I am really glad we have such tests in place. One tiny remark: for your tests could we use a wdio global with 4 underscores `____wdio`? As you can see I added one _ per test to prevent namespace collisions. Maybe it's even a better idea to just use one variable and delete it after each describe block.

So actually this test fails intentional right? Because the master branch of wdio-sync is not published yet or we have an actual bug. Does it pass on your site?

**Comment by georgecrawford** [comment]
@christian-bromann I added tests as promised, but couldn't actually find a failing test case. 

I think these are the rules we now impose on custom commands. Can you check that you agree?

1. custom commands can encapsulate native wdio commands:
```js
        browser.addCommand('customWdio', function (a) {
            browser.pause(1000)
            return a + 1
        })
```
2. custom commands can implement their own routines using Promises (using `async` function name):
```js
        browser.addCommand('customNativePromise', function async (a) {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve(a + 1)
                }, 1000)
            })
        })
```
or:
```js
        browser.addCommand('customQPromise', function async (a) {
            const defer = q.defer()
            setTimeout(() => {
                defer.resolve(a + 1)
            }, 1000)
            return defer.promise
        })
```
3. Custom commands may treat wdio commands as promises (using `async` function name):
```js
        browser.addCommand('customWdioPromise', function async (a) {
            return browser.pause(1000)
            .then(() => {
                return a + 1
            })
        })
```
4. However, within a custom command, calling any other custom command *must* use the sync style:
```js
        // This works:
        browser.addCommand('customWrapWdio', function (a) {
            const b = browser.customWdio(a)
            return b + 1
        })

        // ...but this fails with `Uncaught TypeError: browser.customWdio(...).then is not a function`:
        browser.addCommand('customWrapWdio', function (a) {
            return browser.customWdio(a)
            .then((b) => {
                return b + 1
            })
        })
```

**Comment by georgecrawford** [comment]
Aah no, these tests all pass on `master` of `wdio-sync`

**Comment by christian-bromann** [comment]
```js
        // ...but this fails with `Uncaught TypeError: browser.customWdio(...).then is not a function`:
        browser.addCommand('customWrapWdio', function (a) {
            return browser.customWdio(a)
            .then((b) => {
                return b + 1
            })
        })
```
Here is the `async` function name missing.

**Comment by christian-bromann** [comment]
The rest looks :+1: .. I think this is the best solution we can provide for handling async with WebdriverIO and custom commands. What do you think?

Do you want to publish `wdio-sync` to v0.1.0 and retest this PR to double check?

**Comment by christian-bromann** [comment]
What's your npm account, so I can add you as contributor

**Comment by georgecrawford** [comment]
Re https://github.com/webdriverio/wdio-mocha-framework/pull/6#issuecomment-159669005 - I made a mistake, but it still fails.

**Comment by georgecrawford** [comment]
Re https://github.com/webdriverio/wdio-mocha-framework/pull/6#discussion_r45887909 - removing the lines like `onPrepare: NOOP,` causes big errors:
```
     TypeError: Cannot read property 'map' of undefined
      at executeHooksWithArgs (/Users/george/Sites/Git/wdio-sync/build/index.js:136:18)
      at MochaAdapter.run$ (lib/adapter.js:62:15)
      at tryCatch (node_modules/regenerator/runtime.js:61:40)
```

**Comment by christian-bromann** [comment]
Ok, leave it then.. we will have NOOP set as default in the [config parser](https://github.com/webdriverio/webdriverio/blob/v4.0/lib/utils/ConfigParser.js#L30-L44)

**Pull request merged**